### PR TITLE
[ESQL] Make Datatypes an enum

### PR DESCRIPTION
**Pull request by not-napoleon**
After https://github.com/elastic/elasticsearch/pull/109162, all instances of `DataType` were in one class, as a collection of static constants.  That pattern can then be simplified into an enum, merging the behavior class and the static collection class.  That opens the door to future optimizations, like using enum serialization rather than string serialization to save bytes over the wire. It also makes the code easier to read, as all the behavior is now in a single file (which is still pretty short).

Most of this PR is just juggling names around to have the two references refer to the same thing.  

Follow up work can merge the functions from `EsqlDataTypes` into this enum, but this PR is long enough already.

### CODE REVIEW

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataTypes.java [hunk 1]
```diff
@@ -214,4 +233,54 @@ public static boolean areCompatible(DataType left, DataType right) {
                 || (isDateTime(left) && isDateTime(right));
         }
     }
+
+    public String nameUpper() {
```
Comment (not-napoleon) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataTypes.java:237]: These are the functions from the old `DataType` class.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java [hunk 2]
```diff
@@ -677,10 +676,10 @@ private LogicalPlan resolveEnrich(Enrich enrich, List<Attribute> childrenOutput)
             return enrich;
         }
 
-        private static final DataType[] GEO_TYPES = new DataType[] { GEO_POINT, GEO_SHAPE };
-        private static final DataType[] NON_GEO_TYPES = new DataType[] { KEYWORD, TEXT, IP, LONG, INTEGER, FLOAT, DOUBLE, DATETIME };
+        private static final DataTypes[] GEO_TYPES = new DataTypes[] { GEO_POINT, GEO_SHAPE };
+        private static final DataTypes[] NON_GEO_TYPES = new DataTypes[] { KEYWORD, TEXT, IP, LONG, INTEGER, FLOAT, DOUBLE, DATETIME };
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:680]: Maybe these lists should be pulled up into the enum?  It's also very unclear why `NON_GEO_TYPES` here is only a small subset of the remaining types.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java [hunk 3]
```diff
@@ -897,11 +896,11 @@ private static Expression processIn(In in) {
             return childrenChanged ? in.replaceChildren(newChildren) : in;
         }
 
-        private static boolean supportsImplicitCasting(DataType type) {
+        private static boolean supportsImplicitCasting(DataTypes type) {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:899]: Again, this feels like it should be a method on the enum.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java [hunk 4]
```diff
@@ -43,13 +42,8 @@
...
     // the numeric types convert functions need to handle; the other numeric types are converted upstream to one of these
-    private static final List<DataType> NUMERIC_TYPES = List.of(
-        DataTypes.INTEGER,
-        DataTypes.LONG,
-        DataTypes.UNSIGNED_LONG,
-        DataTypes.DOUBLE
-    );
-    public static final List<DataType> STRING_TYPES = DataTypes.types().stream().filter(EsqlDataTypes::isString).toList();
+    private static final List<DataTypes> NUMERIC_TYPES = List.of(DataTypes.INTEGER, DataTypes.LONG, DataTypes.UNSIGNED_LONG, DataTypes.DOUBLE);
+    public static final List<DataTypes> STRING_TYPES = DataTypes.types().stream().filter(EsqlDataTypes::isString).toList();
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java:46]: These lists also feel like they belong on `DataTypes`.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/ReferenceAttribute.java [hunk 5]
```diff
@@ -8,21 +8,21 @@
...
-import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.core.type.DataTypes;
 
 /**
  * Attribute based on a reference to an expression.
  */
 public class ReferenceAttribute extends TypedAttribute {
 
-    public ReferenceAttribute(Source source, String name, DataType dataType) {
+    public ReferenceAttribute(Source source, String name, DataTypes dataType) {
```
Comment (nik9000) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/ReferenceAttribute.java:18]: I imagine we want the name to be `DataType` not `DataTypes`.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java [hunk 6]
```diff
@@ -541,7 +538,8 @@ public Expression visitFunctionExpression(EsqlBaseParser.FunctionExpressionConte
     @Override
     public Expression visitInlineCast(EsqlBaseParser.InlineCastContext ctx) {
         Source source = source(ctx);
-        DataType dataType = typedParsing(this, ctx.dataType(), DataType.class);
+        // NOCOMMIT this seems like trouble
+        DataTypes dataType = typedParsing(this, ctx.dataType(), DataTypes.class);
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:542]: It's deeply unclear to me why we're passing in `DataType.class` here; It reads like we're checking that the type parsing returns a type? but that doesn't make sense, so I assume these undocumented methods are doing something non-obvious.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexResolver.java [hunk 7]
```diff
@@ -25,14 +25,8 @@
 import org.elasticsearch.index.mapper.TimeSeriesParams;
 import org.elasticsearch.transport.NoSuchRemoteClusterException;
 import org.elasticsearch.xpack.esql.core.QlIllegalArgumentException;
+import org.elasticsearch.xpack.esql.core.type.*;
```
Comment (nik9000) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexResolver.java:28]: That's going to blow up I think.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexCompatibility.java [hunk 8]
```diff
@@ -27,9 +27,9 @@ public static Map<String, EsField> compatible(Map<String, EsField> mapping, Vers
             if (isPrimitive(dataType) == false) {
                 compatible(esField.getProperties(), version);
             } else if (isTypeSupportedInVersion(dataType, version) == false) {
-                EsField field = new UnsupportedEsField(entry.getKey(), dataType.name(), null, esField.getProperties());
+                EsField field = new UnsupportedEsField(entry.getKey(), dataType.nameUpper(), null, esField.getProperties());
                 entry.setValue(field);
-                propagateUnsupportedType(entry.getKey(), dataType.name(), esField.getProperties());
+                propagateUnsupportedType(entry.getKey(), dataType.nameUpper(), esField.getProperties());
```
Comment (nik9000) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexCompatibility.java:32]: Is this meant to change?

**Review by nik9000 — CHANGES_REQUESTED** [review]
I think this is something we should do. We're not going to allow plugging in new data types any time soon and this seems like a nicer way to organize the types in the mean time.

I think I found a bug. Or at least something to talk about.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexResolver.java (hunk 7 above)
Comment (not-napoleon) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexResolver.java:28]: ah, crumb, I hate it when intellij does that.  and with the NOCOMMIT line still in here, precommit didn't get to the step that flags this.  Thanks for spotting it.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java [hunk 9]
```diff
@@ -56,7 +102,139 @@ public DataType(String typeName, String esType, int size, boolean isInteger, boo
...
+        DATE_PERIOD,
+        TIME_DURATION,
+        GEO_POINT,
+        CARTESIAN_POINT,
+        CARTESIAN_SHAPE,
+        GEO_SHAPE,
+        COUNTER_LONG,
+        COUNTER_INTEGER,
+        COUNTER_DOUBLE
+    ).sorted(Comparator.comparing(DataType::typeName)).toList();
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:135]: It is not all of them.  It doesn't include `DOC_DATA_TYPE` and `TSID_DATA_TYPE`.  Maybe it should, but it didn't when I refactored this.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexCompatibility.java (hunk 8 above)
Comment (not-napoleon) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexCompatibility.java:32]: I think so?  This is the one tricky bit with this refactor, actually.  The old `DataType` class defined a `name` method, which is not the same as the enum name method, and you can't override enum's name.  You can see [here](https://github.com/elastic/elasticsearch/pull/109227/files#diff-0635d6d9ed5349bb1eeff14286503a11cce10e57bc4fc18f6ba2e35c0fa11bfeL49) where we generate the name as the upper case of either the type name or the es type (and I don't know why those are different; reconciling the two constructors for this class would be an excellent follow-up).  So when I had to rename the function, I thought `nameUpper` made some sense.

In a perfect world, the constant name, `typeName`, and `esType` would all be the same thing, and we could just use `DataType::name`, but that's not what it is right now.  It looks like the only ones where the results of `name` are different from the old results of `name` are `SOURCE`, `DOC_DATA_TYPE`, and `TSID_DATA_TYPE` (which would be `_SOURCE`, `_DOC`, and `_TSID`, respectively).  If we don't care about those, or don't mind renaming the constants, then we can replace this with the enum version of `name` and not stress it.

So unless something more subtle than my (currently pre-coffee) brain can pick up on is happening here, this is intentional, but for a bad reason.  I'm open to fixing it, but didn't want to just start changing behavior in what's supposed to be a pure refactor.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:135]: In that case it'd be more clear *to me* if we iterated all of them and explicitly subtracted those two. Certainly not a blocker, but an option.

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexCompatibility.java (hunk 8 above)
Comment (nik9000) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/index/IndexCompatibility.java:32]: gotcha

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:135]: I'm happy to do that as a follow up PR, where it'll be clearer what's being changed.

[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:135]: :metal:

**Review by nik9000 — APPROVED** [review]

**Review by nik9000 — APPROVED** [review]
I have reviewed, with my own personal eyes, all the changes in this file. It's `s/DataTypes/DataType/`  with a few `nameUpper` and the actual move of the data types. I see nothing else. I want this.

**Pull request merged**